server/exporter.js

```diff
@@ -15,45 +19,188 @@ const queueOptions = {
   },
 };
 
-async function exportAndEmailCollections(teamId: string, email: string) {
+async function fileOperationsUpdate(teamId, userId, exportData) {
+  await Event.add({
+    name: "fileOperations.update",
+    teamId: teamId,
+    actorId: userId,
+    data: {
+      type: exportData.type,
+      id: exportData.id,
+      state: exportData.state,
+      size: exportData.size,
+      collectionId: exportData.collectionId,
+      createdAt: exportData.createdAt,
+    },
+  });
+}
+
+type exportAndEmailCollectionsType = {|
+  teamId: string,
+  userId: string,
+  email: string,
+  collectionId?: string,
+  fileOperationId?: string,
+|};
+
+async function exportAndEmailCollections({
+  teamId,
+  userId,
+  email,
+  collectionId,
+  fileOperationId,
+}: exportAndEmailCollectionsType) {
   log("Archiving team", teamId);
   const { archiveCollections } = require("./utils/zip");
   const team = await Team.findByPk(teamId);
-  const collections = await Collection.findAll({
-    where: { teamId },
-    order: [["name", "ASC"]],
-  });
+  const user = await User.findByPk(userId);
+
+  let collections;
+  if (!collectionId) {
+    collections = await Collection.scope({
+      method: ["withMembership", userId],
+    }).findAll({
+      where: { teamId },
+      order: [["name", "ASC"]],
+    });
+
+    collections = collections.reduce((agg, collection) => {
+      if (can(user, "read", collection)) {
+        return [...agg, collection];
+      }
+      return agg;
+    }, []);
+  } else {
+    collections = [await Collection.findByPk(collectionId)];
+  }
+
+  const acl = process.env.AWS_S3_ACL || "private";
+
+  let exportData;
+  let state;
+  let key;
+
+  if (fileOperationId) {
```

Comment:
I'm trying to follow why in one case the fileOperationId is created outside and passed in, but in the other it's inside of this method, are you able to explain?

Overall everything works well but this method feels "messy"

Reply:
Modified both the endpoints to create fileOperation before putting tasks on the queue